### PR TITLE
Add MultiNamedReservableCurrency

### LINKED ISSUE #668: NamedReservableCurrency equivalent for MultiReservableCurrency
Just as with [`NamedReservableCurrency`](https://github.com/paritytech/substrate/blob/master/frame/support/src/traits/tokens/currency/reservable.rs#L84), it is quite useful if the same can exist for the `MultiCurrency` family. Pallets can reserve under a namespace without fear of interfering with other modules.

### PR DESCRIPTION
**Pull request by stanly-johnson**
resolves #668 

- adds MultiNamedReservableCurrency trait
- impls MultiNamedReservableCurrency trait in orml-tokens
- impls NamedReservableCurrency in orml-tokens

### CODE REVIEW

File: tokens/src/lib.rs [hunk 1]
```diff
@@ -1276,6 +1314,253 @@ impl<T: Config> MultiReservableCurrency<T::AccountId> for Pallet<T> {
...
+	/// according to `status`. If `status` is `Reserved`, the balance will be
+	/// reserved with given `id`.
+	///
+	/// Is a no-op if:
+	/// - the value to be moved is zero; or
+	/// - the `slashed` id equal to `beneficiary` and the `status` is
+	///   `Reserved`.
+	fn repatriate_reserved_named(
+		id: &Self::ReserveIdentifier,
+		currency_id: T::CurrencyId,
```
Comment (zqhxuyuan) [inline tokens/src/lib.rs:1460]: change to `Self::CurrencyId`?

File: tokens/src/lib.rs [hunk 2]
```diff
@@ -225,6 +242,8 @@ pub mod module {
 		ExistentialDeposit,
 		/// Beneficiary account must pre-exist
 		DeadAccount,
+		/// Number of named reserves exceed MaxReserves
```
Comment (shaunxw) [inline tokens/src/lib.rs:245]:
```suggestion
		/// Number of named reserves exceed `T::MaxReserves`
```

File: tokens/src/lib.rs [hunk 3]
```diff
@@ -278,6 +297,12 @@ pub mod module {
 			free: T::Balance,
 			reserved: T::Balance,
 		},
+		/// Some amount was removed from the account (e.g. for misbehavior).
+		Slashed {
+			currency_id: T::CurrencyId,
+			who: T::AccountId,
+			amount: T::Balance,
+		},
```
Comment (shaunxw) [inline tokens/src/lib.rs:305]: would be conflict with `Slashed` event in https://github.com/open-web3-stack/open-runtime-module-library/pull/729

File: tokens/src/lib.rs [hunk 4]
```diff
@@ -1276,6 +1314,253 @@ impl<T: Config> MultiReservableCurrency<T::AccountId> for Pallet<T> {
...
+					reserves[index].amount = reserves[index].amount.defensive_saturating_add(value);
+				}
+				Err(index) => {
+					reserves
+						.try_insert(index, ReserveData { id: *id, amount: value })
+						.map_err(|_| Error::<T>::TooManyReserves)?;
+				}
+			};
+			<Self as MultiReservableCurrency<_>>::reserve(currency_id, who, value)?;
+			Ok(())
```
Comment (shaunxw) [inline tokens/src/lib.rs:1359]:
```suggestion
			<Self as MultiReservableCurrency<_>>::reserve(currency_id, who, value)
```

**Review by shaunxw — APPROVED** [review]

**Pull request merged**